src/main/groovy/pl/allegro/tech/build/axion/release/domain/NextVersionMarker.groovy

```diff
@@ -9,22 +10,34 @@ class NextVersionMarker {
 
     private static final ReleaseLogger logger = ReleaseLogger.Factory.logger(NextVersionMarker)
 
-    private final ScmService repository
+    private final ScmService repositoryService
+    private final VersionConfig versionConfig
 
-    NextVersionMarker(ScmService repository) {
-        this.repository = repository
+    NextVersionMarker(ScmService repositoryService, VersionConfig config) {
+        this.repositoryService = repositoryService
+        this.versionConfig = config
     }
 
     void markNextVersion(NextVersionProperties nextVersionRules, TagProperties tagRules) {
-        if (nextVersionRules.nextVersion == null) {
-            throw new IllegalArgumentException("No next version specified! Use -Prelease.version to set next version.")
+
+        def version = nextVersionRules.nextVersion
+        def currentVersion = Version.valueOf(versionConfig.version)
+
+        if (nextVersionRules.versionIncrementer) {
```

Comment:
`NextVersionPropertiesFactory` is a place where you should do `nextVersionRules.versionIncrementer` resolution. If you need `VersionConfig` in `NextVersionPropertiesFactory`, just change the signature.

This way code in all services is cleaner, as input values are already resolved.